### ISSUE TITLE
Some questions regarding V2

### ISSUE THREAD

**Issue by michaelbrusegard** [vantezzen/autoform#122]
1. What is the point of the ErrorMessage field component now that it is handled inside the FieldWrapper?
2. How do I add my custom `topContent` prop to the fieldconfig? I get the type error because the FieldWrapperProps type does not know about it when i update the fieldconfig type.
3. In V1 it showed an error message when the input fields were empty because I have supplied and required error message in the zodschema, but in V2 it does not show it only focuses the input field without showing the message.
4. The type error below when passing in formComponents as a prop on AutoForm as shown in the example:
<img width="1027" alt="Screenshot 2024-10-22 at 15 34 10" src="https://github.com/user-attachments/assets/c4a1f6b8-8835-40be-926a-dde7f1aaf3b9">
5. The type error below when passing my most complex password input schema to the zodprovider:
<img width="1661" alt="Screenshot 2024-10-22 at 15 35 39" src="https://github.com/user-attachments/assets/fa444d8e-905e-49ee-b089-c8dafe41da48">

Here is the schema in question:
```tsx
function signUpPasswordSchema(
  t?: (key: string) => string,
  defaultValues?: { email: string },
) {
  let schema = z.object({});

  if (defaultValues?.email) {
    schema = schema.extend({
      email: z
        .string()
        .optional()
        .default(defaultValues?.email as unknown as string)
        .superRefine(
          fieldConfig({
            fieldType: 'hidden',
          }),
        ),
    });
  }

  schema = schema.extend({
    password: z
      .string({
        required_error: t
          ? t('form.password.required')
          : 'Password is required',
      })
      .min(8, {
        message: t
          ? t('form.password.minLength')
          : 'Password must be at least 8 characters long',
      })
      .max(50, {
        message: t
          ? t('form.password.maxLength')
          : 'Password must be at most 50 characters long',
      })
      .regex(/[A-Z]/, {
        message: t
          ? t('form.password.uppercase')
          : 'Password must contain at least one uppercase letter',
      })
      .regex(/[^a-zA-Z0-9]/, {
        message: t
          ? t('form.password.specialChar')
          : 'Password must contain at least one special character',
      })
      .describe(t ? t('form.password.label') : 'Password')
      .superRefine(
        fieldConfig({
          fieldType: 'password',
          inputProps: {
            autocomplete: 'new-password',
          },
        }),
      ),
  });

  if (t) {
    const schemaWithConfirm = schema
      .extend({
        confirmPassword: z
          .string({
            required_error: t('form.password.mismatch'),
          })
          .describe(t('form.password.confirmlabel'))
          .superRefine(
            fieldConfig({
              fieldType: 'password',
              inputProps: {
                autocomplete: 'new-password',
              },
            }),
          ),
      })
      .refine((data) => data.password === data.confirmPassword, {
        message: t('form.password.mismatch'),
        path: ['confirmPassword'],
      });
    return schemaWithConfirm;
  }

  return schema;
}
```

I will update the issue if I encounter any new errors in V2.

**Comment by michaelbrusegard** [comment]
I just found out that defaultValues are now available as a prop on the AutoForm object so I do not have to pass them through the schema :)

**Comment by michaelbrusegard** [comment]
```tsx
import type { AutoFormFieldProps } from '@autoform/react';

import { cx } from '@/lib/utils';

import { DatePicker } from '@/components/composites/DatePicker';

const DateField: React.FC<AutoFormFieldProps> = ({
  field,
  inputProps,
  error,
  id,
}) => {
  return (
    <DatePicker
      id={id}
      className={cx(error && 'border-destructive')}
      date={inputProps.value}
      setDate={(date) => {
        const event = {
          target: {
            name: field.key,
            value: date,
          },
        };
        inputProps.onChange(event);
      }}
      {...inputProps}
    />
  );
};

export { DateField };
```

I tried modifying the DateField to work with V2, but I am unable to set its value

**Comment by vantezzen** [comment]
> What is the point of the ErrorMessage field component now that it is handled inside the FieldWrapper?

I guess not much now. It was originally meant to be used for all validations and as a way for AutoForm to present developer errors but apart from "I can't find a field declaration for that field type" there aren't any rendering errors so far. I'll keep it in the declaration for now but if no further use is found I might remove it if it's really useless.

> How do I add my custom topContent prop to the fieldconfig? I get the type error because the FieldWrapperProps type does not know about it when i update the fieldconfig type

They are under the `customData` key of the fieldConfig so that future API changes don't conflict with custom data. You should be able to supply and use it with the `buildZodFieldProps` function.

> In V1 it showed an error message when the input fields were empty because I have supplied and required error message in the zodschema, but in V2 it does not show it only focuses the input field without showing the message.

This might be a thing because Zod allows empty strings to satisfy z.string() types? If the field has not been touched and no default value (including "") has been supplied, the value is undefined internally so it fails but as soon as an empty string is set, the string is allowed.

> The type error below when passing in formComponents as a prop on AutoForm as shown in the example

Where does that "| undefined" come from? As far as I can see in the props, there shouldn't be an optional

> The type error below when passing my most complex password input schema to the zodprovider

I think this might be an issue with how TypeScript tries to determine the return value of that function? If I try to use z.infer on that definition, it comes back as `{}`. Otherwise I see that it shows some error with the ZodEffect, even though this should be fixed with 2.1.0 so I might have to fix something with nested Effects.

> I tried modifying the DateField to work with V2, but I am unable to set its value

react-hook-form expects the value to be a string which might be the issue here. You can try using `value: date.toISOString()` - the date coersion is done later via the schema.
